spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,223 @@
+    if (endOffset == null) {
+      throw new IllegalStateException("End offset is not set");
+    }
+
+    return endOffset;
+  }
+
+  @Override
+  public Offset deserializeOffset(String json) {
+    return StreamingOffset.fromJson(json);
+  }
+
+  @Override
+  public void commit(Offset end) {}
+
+  @Override
+  public void stop() {}
+
+  @Override
+  protected List<CombinedScanTask> tasks() {
+    Preconditions.checkState(startOffset != null,
+        "Start offset is null when planning tasks, this probably because there's no snapshot in the table currently");
+    Preconditions.checkState(endOffset != null, "End offset is null when planning tasks");
+
+    StreamingOffset start = (StreamingOffset) startOffset;
+    StreamingOffset end = (StreamingOffset) endOffset;
+
+    List<SnapshotTasks> snapshotTasksList = getChanges(start.snapshotId(), start.index(), start.isStartingSnapshotId());
+    List<CombinedScanTask> runningTasks = Lists.newArrayList();
+    for (SnapshotTasks s : snapshotTasksList) {
+      if (s.snapshotId() == end.snapshotId()) {
+        runningTasks.addAll(s.tasks().subList(0, end.index() + 1));
+        break;
+      } else {
+        runningTasks.addAll(s.tasks());
+      }
+    }
+
+    return runningTasks;
+  }
+
+  private StreamingOffset calculateInitialOffset() {
+    Snapshot current = table.currentSnapshot();
+    List<Long> snapshotIds = SnapshotUtil.ancestorIds(current, table::snapshot);
+    if (snapshotIds.isEmpty()) {
+      return null;
+    } else {
+      long initSnapshotId = snapshotIds.get(snapshotIds.size() - 1);
+      return new StreamingOffset(initSnapshotId, -1, true /* is_starting */);
+    }
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    List<SnapshotTasks> snapshotTasksList = getChangesWithRateLimit(start.snapshotId(),
+        start.index(), start.isStartingSnapshotId());
+    SnapshotTasks last = snapshotTasksList.isEmpty() ? null : snapshotTasksList.get(snapshotTasksList.size() - 1);
+
+    return last == null ? start : new StreamingOffset(last.snapshotId, last.size() - 1,
+        last.snapshotId() == start.snapshotId() /* is_starting */);
+  }
+
+  @VisibleForTesting
+  List<SnapshotTasks> getChanges(long snapshotId, int index, boolean isStarting) {
+    List<SnapshotTasks> snapshotTasksList = Lists.newArrayList();
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+
+    if (isStarting) {
+      CloseableIterable<CombinedScanTask> iter = buildTableScan().useSnapshot(snapshotId).planTasks();
+      SnapshotTasks snapshotTasks = new SnapshotTasks(snapshotId, Lists.newArrayList(iter));
+      snapshotTasksList.add(snapshotTasks);
+    }
+
+    Long startSnapshotId = isStarting ? Long.valueOf(snapshotId) : table.snapshot(snapshotId).parentId();
+    Preconditions.checkState(startSnapshotId != null, "Start snapshot id should be existed");
+    List<Long> snapshotIds = SnapshotUtil.snapshotIdsBetween(table, startSnapshotId, currentSnapshotId);
```

Comment:
This method doesn't check that `startSnapshotId` is an ancestor of `currentSnapshotId`. We should check that the starting snapshot is an ancestor of the current one, or else we need to throw an exception. Feel free to update `SnapshotUtil` to throw in `snapshotIdsBetween`, I think it would make sense for that method to handle it.

Reply:
Done.